teps/add-variable-retries-and-retry-count.md

```diff
@@ -0,0 +1,105 @@
+---
+title: add-variable-retries-and-retry-count
+authors:
+  - "@zhangyumeng"
+creation-date: 2020-10-14
+last-updated: 2020-10-19
+status: proposed
+---
+
+# TEP-0026: Add Variable `retries` and `retry-count`
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories)
+- [Alternatives](#Alternatives)
+- [References](#references-optional)
+<!-- /toc -->
+
+## Summary
+
+This TEP proposes adding `retries` and current retry times to Pipeline variables.
+
+## Motivation
+
+PipelineTask can be rerun after failure when user specifies `retries` in PipelineTask. In that case, exposing the retry count can give user more information. User can dispatch a notification after certain times of failure, or customize logs according to retry count.
+
+### Goals
+
+- add 2 variables indicating retries and retry count of a PipelineTask
+
+### Non-Goals
+
+- execute different PipelineTask based on retry count
+- emit an event on every retry
+
+## Proposal
+
+There are 2 variable substitutions `tasks.<taskName>.retries` and `tasks.<taskName>.retry-count` needed to be supported:
+
+1. For retrieving `retries`, we can use `Retries` that stored in `PipelineTask` struct
+2. For retrieving `retry-count`, we can use length of `RetriesStatus` in `TaskRunStatus`
+
+Code example:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Task
+metadata:
+  name: failed-task
+spec:
+  params:
+    - name: num-retries
+      type: string
+    - name: max-retries
+      type: string
+  steps:
+    - image: alpine:3.12.0
+      script: |
+        #!/usr/bin/env sh
+        if [ $(params.num-retries) == $(params.max-retries) ]; then
+          echo "This is the last retry. \n"
+        fi
+    - image: ubuntu
+      script: |
+        #!/usr/bin/env sh
+        /path/to/<program-that-fails>
+```
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+metadata:
+  name: retry-example
+spec:
+  tasks:
+  - name: retry-me
+    retries: 5
+    taskRef:
+      name: failed-task
+    params:
+    - name: num-retries
+      value: $(tasks.retry-me.retry-count)
+    - name: max-retries
+      value: $(tasks.retry-me.retries)
+```
+
+In this example, Tekton will echo `This is the last retry` when retry count is equal to 5.
+
+### User Stories
+
+`PipelineTask` author wants to send a Slack notification only if it fails after specified `retries`, not every time it fails.The user wants less interruption.
```

Comment:
Is this the only use case we are envision for this ? 
Is there any other potential use for this feature ? On top of my head, I could see a need to be able to parametrize the task with the retry-count, for example, if the task provision something (a cluster, a VM, …) and want to name it accordingly to the retry-count (in order to keep the previous one around or something along those lines) — this particular use case could be done using different variables though (using the uid, …).